### PR TITLE
[Merged by Bors] - feat(Tactic/ByContra): by_contra' tactic

### PR DESCRIPTION
**Pull request by kbuzzard**
Implementation of `by_contra'`. Note that it needs the (as yet unimplemented) `push_neg`.

### CODE REVIEW

File: test/byContra.lean [hunk 1]
```diff
@@ -0,0 +1,35 @@
+-- tests for byContra' tactic
+import Mathlib.Tactic.ByContra
+
+-- Basic implementation of `push_neg` to make it look like the tactic's working
+-- This needs to be removed when `push_neg` is actually implemented.
+macro_rules
+  | `(tactic| push_neg) => `(tactic| simp only [not_not, not_lt])
+  | `(tactic| push_neg at $e) => `(tactic| simp only [not_not, not_lt] at $e)
```
Comment (digama0) [inline test/byContra.lean:8]:
```suggestion
  | `(tactic| push_neg $(loc)?) => `(tactic| simp only [not_not, not_lt] $(loc)?)
```

**Comment by kbuzzard** [comment]
@dupuisf and Jireh Loreaux are porting `push_neg` and I think that it would be simpler to wait until that is done before merging this (I'll have to fix the tests anyway).

**Comment by kbuzzard** [comment]
The `push_neg` PR is #344.

**Comment by kbuzzard** [comment]
I tried to get this working, now push_neg has been merged, but apparently my code doesn't work any more, I don't know why, and unfortunately I don't have the time to think about this right now (I wrote this code in July whilst in the same room as Leo and since then have forgotten what little I knew about tactic writing).

**Comment by gebner** [comment]
Congratulations!  You've found an exciting bug in `push_neg`!  It was the usual missing `instantiateMVars`.

There's still one failing test, I'll let you fix it.

BTW, `guard_hyp foo : ty` works, only `guard_hyp this : ty` doesn't.  The bug fix is in https://github.com/leanprover/std4/commit/7787697470956ebf7a74c6b76b387d2da08d0b5f but we still need to import std4 here.

**Comment by gebner** [comment]
bors merge

**Comment by kbuzzard** [comment]
Oh thanks for the explanation and the fix Gabriel! I'm sorry I didn't finish the job, I don't usually look much at github and I only just now saw the notification.

**Pull request closed without merging**